File: Merge Sort/MergeSort.cpp
#include <iostream>
#include <stdlib.h>
#include <time.h>
#include <vector>


std::vector<int> merge_sort(std::vector<int> arr)
{
	int length = arr.size();

	if (length == 1)
	{
		return arr;
	}

	int middle = length / 2;

	std::vector<int> sorted_left = merge_sort(std::vector<int>(arr.begin(), arr.begin() + middle));
	std::vector<int> sorted_right = merge_sort(std::vector<int>(arr.begin() + middle, arr.end()));

	int left_length = sorted_left.size();
	int right_length = sorted_right.size();

	int left_pointer = 0, right_pointer = 0;

	std::vector<int> temp_vector;

	while (left_pointer < left_length && right_pointer < right_length)
	{
		int left_value = sorted_left[left_pointer];
		int right_value = sorted_right[right_pointer];

		if (left_value < right_value)
		{
			temp_vector.push_back(left_value);
			left_pointer += 1;
		}
		else
		{
			temp_vector.push_back(right_value);
			right_pointer += 1;
		}
	}

	while (left_pointer < left_length)
	{
		temp_vector.push_back(sorted_left.at(left_pointer));
		left_pointer += 1;
	}

	while (right_pointer < right_length)
	{
		temp_vector.push_back(sorted_right.at(right_pointer));
		right_pointer += 1;
	}

	return temp_vector;
}


int main() 
{
	int arr_length = 1000000;

	std::vector<int> arr(arr_length);

	srand(time(0));
	
	// Add random numbers to array
	for (int i = 0; i < arr_length; i++)
	{
		// Get number in range 0 - 100;
		int num = (rand() % (arr_length - 0 + 1)) + 0;
		arr.at(i) = num;
		//std::cout << arr.at(i) << " ";
	}
	
	clock_t t;
	t = clock();

	std::cout << "Sorting " << arr_length << " data...\n\n";

	std::vector<int> sorted_array = merge_sort(arr);
	
	t = clock() - t;
	double time_taken = ((double)t) / CLOCKS_PER_SEC; // calculate the elapsed time
	printf("Sorting took %f seconds to complete ", time_taken);

	std::cout << sorted_array.size() << " data\n\n";

	return 0;
}

File: Merge Sort/MergeSort.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class MergeSort {

    public static List<Integer> mergeSort(List<Integer> list) {
        int middle = list.size() / 2;

        List<Integer> leftArray = list.subList(0, middle);
        List<Integer> rightArray = list.subList(middle, list.size());

        List<Integer> sortedLeft = leftArray;
        List<Integer> sortedRight = rightArray;

        if (leftArray.size() > 1)
            sortedLeft = mergeSort(leftArray);

        if (rightArray.size() > 1)
            sortedRight = mergeSort(rightArray);

        int leftPointer = 0;
        int rightPointer = 0;

        // merge
        List<Integer> tempMergedArray = new ArrayList<>();
        while (leftPointer < sortedLeft.size() && rightPointer < sortedRight.size()) {
            int leftValue = sortedLeft.get(leftPointer);
            int rightValue = sortedRight.get(rightPointer);

            if (leftValue < rightValue) {
                tempMergedArray.add(leftValue);
                leftPointer += 1;
            }
            else {
                tempMergedArray.add(rightValue);
                rightPointer += 1;
            }
        }

        while (leftPointer < sortedLeft.size()) {
            tempMergedArray.add(sortedLeft.get(leftPointer));
            leftPointer += 1;
        }

        while (rightPointer < sortedRight.size()) {
            tempMergedArray.add(sortedRight.get(rightPointer));
            rightPointer += 1;
        }

        return tempMergedArray;
    }

    public static void main(String[] args) {
        List<Integer> arr = new ArrayList<>();

        int nData = 25;

        Random rand = new Random();
        // Add random numbers to array list
        for (int i = 0; i < nData; i++) {
            // Generate random number from 0 to 99;
            int number = rand.nextInt(100);
            arr.add(number);
            System.out.print(arr.get(i) + " ");
        }
        System.out.println();

        List<Integer> sortedArray = mergeSort(arr);

        System.out.println("SortedArray:");
        for (int i = 0; i < sortedArray.size(); i++) {
            System.out.print(sortedArray.get(i) + " ");
        }
    }
}
